v2/apiserver/internal/core/events.go

```diff
@@ -403,13 +419,32 @@ func (e *eventsService) createSingleEvent(
 
 	event.ID = uuid.NewV4().String()
 
+	jobs := []Job{}
 	workerSpec := project.Spec.WorkerTemplate
+	// If the event is a retry of another, defer to the Worker.Spec on the event
+	// itself.  Retain/cache any successful jobs that do not use a shared
+	// workspace.
+	if event.Labels != nil && event.Labels[RetryLabelKey] != "" {
+		workerSpec = event.Worker.Spec
+		for _, job := range event.Worker.Jobs {
+			if job.Status.Phase == JobPhaseSucceeded && !job.UsesWorkspace() {
+				// Capture event ID for tracing original logs.  Note that it may
+				// already be set (e.g. via a retry of another retry), in which case
+				// we do not want to override it.
+				if job.Status.LogsEventID == "" {
+					job.Status.LogsEventID = event.Labels[RetryLabelKey]
+				}
+				jobs = append(jobs, job)
+			}
+		}
+	}
```

Comment:
I'm wondering if it may end up feeling cleaner to move retry-specific code into the `retry()` function instead of conditionally doing this stuff `createSingleEvent()` when it's determined that we're dealing with a retry... or you may have had reasons for doing it this way...

Reply:
Good call.  Job inheritance logic moved to Retry.